### PR TITLE
feat: Add Pancake Sort algorithm

### PR DESCRIPTION
**Pull request by divyansh12323**
#### Checklist

- [x ] Added description of change
- [x ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x ] Added tests and example, test must pass
- [x ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x ] Relevant documentation/comments is changed or added
- [x ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x ] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1189"></a>

### CODE REVIEW

File: sorting/pancake_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	                mi = j;
+	            }
+	        }
+		    if (mi != i-1) //check for reversing 
+			 { 
+			    reverse(arr, 0, mi); 
+			    reverse(arr, 0, i-1); 
+		    } 
+	    } 
+    }
```
Comment (ridhishjain) [inline sorting/pancake_sort.cpp:64]: @divyansh12323 the return type is int you haven't returned anything.
That's why one of the checks is failing

**Review by ridhishjain — CHANGES_REQUESTED** [review]

File: sorting/pancake_sort.cpp (hunk 1 above)
Comment (divyansh12323) [reply sorting/pancake_sort.cpp:64]: got it , i'll change

File: sorting/pancake_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    std::cout<<"\nTest 3- as std::vector<float>...";
+    std::vector <float> arr3 = {6.56, 12.62, 200.78, 768.484, 19.27, 68.87, 9.6};
+    sorting::pancakeSort(arr3, size3);
+    assert(std::is_sorted(arr3.begin(), arr3.end()));
+    std::cout<<"Passed\n";
+    for (int i = 0; i < size3; i++) {
+        std::cout<<arr3[i]<<", ";
+    }
+    std::cout<<std::endl;
+
```
Comment (rishabh-997) [inline sorting/pancake_sort.cpp:108]: remove this extra line

File: sorting/pancake_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		    arr[end] = temp; 
+		    start++; 
+		    end--; 
+	    } 
+    } 
+
+    template <typename T>
+    int pancakeSort(std::vector <T>& arr, int n)  {
+        for (int i=n; i>1; --i) {
+            int mi = 0, j; //intialize some variables.
```
Comment (rishabh-997) [inline sorting/pancake_sort.cpp:50]: use some more meaningful naming conventions instead of mi

File: sorting/pancake_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    }
+    std::cout<<std::endl;
+    
+    // example 2: vector of double
+    const int size2 = 8;
+    std::cout<<"\nTest 2- as std::vector<double>...";
+    std::vector <double> arr2 = {23.56, 10.62, 200.78, 111.484,3.9, 1.2, 61.77, 79.6};
+    sorting::pancakeSort(arr2, size2);
+    assert(std::is_sorted(arr2.begin(), arr2.end()));
+    std::cout<<"Passed\n";
```
Comment (rishabh-997) [inline sorting/pancake_sort.cpp:91]: correct spacing in all these lines like this
`std::cout << "Passed\n"`

**Review by rishabh-997 — CHANGES_REQUESTED** [review]
Can you look into these comments... Would love to have your thoughts as well @Panquesito7

File: sorting/pancake_sort.cpp (hunk 3 above)
Comment (divyansh12323) [reply sorting/pancake_sort.cpp:50]: ok i will

File: sorting/pancake_sort.cpp (hunk 4 above)
Comment (divyansh12323) [reply sorting/pancake_sort.cpp:91]: sure

File: sorting/pancake_sort.cpp (hunk 2 above)
Comment (divyansh12323) [reply sorting/pancake_sort.cpp:108]: yes , i w.ill do so.

**Comment by divyansh12323** [comment]
I would like to know your comments as well @kvedala

File: sorting/pancake_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ * Overall time complexity of pancake sort is O(n^2)
+ * For example: example 1:-
+ * Disordered pancake sizes: {2,5,3,7,8}
+ * Sorted: {2,3,5,7,8}
+ * For example: example 2:-
+ * Disordered pancake sizes: {22,51,37,73,81}
+ * Sorted: {22,37,51,73,81}
+ * @author [Divyansh Gupta](https://github.com/divyansh12323)
+ * @see more on [Pancake sort](https://en.wikipedia.org/wiki/Pancake_sorting)
+ * @see related problem [Leetcode](https://leetcode.com/problems/pancake-sorting/)
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:17]:
```suggestion
 * @see related problem at [Leetcode](https://leetcode.com/problems/pancake-sorting/)
```

File: sorting/pancake_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    sorting::pancakeSort(arr3, size3);
+    assert(std::is_sorted(arr3.begin(), arr3.end()));
+    std::cout << "Passed\n";
+    for (int i = 0; i < size3; i++) {
+        std::cout << arr3[i] << ", ";
+    }
+    std::cout << std::endl;
+}
+/**
+ * Our main function with example of sort method.
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:110]:
```suggestion
 * @brief Main function
 * @returns 0 on exit
```

File: sorting/pancake_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                reverse(arr, 0, max_index);
+                reverse(arr, 0, i - 1);
+            }
+        }
+        return 0;
+    }
+} // namespace sorting
+
+/**
+ * Test function
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:69]:
```suggestion
 * @brief Test implementations
 * @returns void
```

File: sorting/pancake_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+*/
+
+#include <iostream>   // for io operations
+#include <vector>     // for std::vector
+#include <algorithm>  // for std::is_sorted
+#include <cassert>    // for std::assert
+
+/**
+ * @namespace sorting
+ * Sorting algorithms
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:27]:
```suggestion
 * @brief Sorting algorithms
```

File: sorting/pancake_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            temp = arr[start];
+            arr[start] = arr[end];
+            arr[end] = temp;
+            start++;
+            end--;
+        }
+    }
+
+    template<typename T>
+    int pancakeSort(std::vector<T> &arr, int size) {
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:48]: Missing function documentation.

File: sorting/pancake_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+#include <algorithm>  // for std::is_sorted
+#include <cassert>    // for std::assert
+
+/**
+ * @namespace sorting
+ * Sorting algorithms
+ */
+namespace sorting {
+    /**
+    * This implementation is for a C-style array input that gets modified in place.
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:31]:
```suggestion
    * @brief This implementation is for a C-style array input that gets modified in place.
```

File: sorting/pancake_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+#include <iostream>   // for io operations
+#include <vector>     // for std::vector
+#include <algorithm>  // for std::is_sorted
+#include <cassert>    // for std::assert
+
+/**
+ * @namespace sorting
+ * Sorting algorithms
+ */
+namespace sorting {
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:29]:
```suggestion
namespace sorting {
/**
 * @namespace pancake_sort
 * @brief Functions for [Pancake sort](https://en.wikipedia.org/wiki/Pancake_sorting) algorithm
 */
namespace pancake_sort {
```

File: sorting/pancake_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            }
+            if (max_index != i - 1) //check for reversing
+            {
+                reverse(arr, 0, max_index);
+                reverse(arr, 0, i - 1);
+            }
+        }
+        return 0;
+    }
+} // namespace sorting
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:66]:
```suggestion
}  // namespace pancake_sort
}  // namespace sorting
```

File: sorting/pancake_sort.cpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * Test function
+ */
+static void test() {
+
+    // example 1: vector ofi nt
+    const int size1 = 7;
+    std::cout << "\nTest 1- as std::vector<int>...";
+    std::vector<int> arr1 = {23, 10, 20, 11, 12, 6, 7};
+    sorting::pancakeSort(arr1, size1);
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:77]:
```suggestion
    sorting::pancake_sort::pancakeSort(arr1, size1);
```

File: sorting/pancake_sort.cpp [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    for (int i = 0; i < size1; i++) {
+        std::cout << arr1[i] << " ,";
+    }
+    std::cout << std::endl;
+
+    // example 2: vector of double
+    const int size2 = 8;
+    std::cout << "\nTest 2- as std::vector<double>...";
+    std::vector<double> arr2 = {23.56, 10.62, 200.78, 111.484, 3.9, 1.2, 61.77, 79.6};
+    sorting::pancakeSort(arr2, size2);
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:89]:
```suggestion
    sorting::pancake_sort::pancakeSort(arr2, size2);
```

File: sorting/pancake_sort.cpp [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    for (int i = 0; i < size2; i++) {
+        std::cout << arr2[i] << ", ";
+    }
+    std::cout << std::endl;
+
+    // example 3:vector of float
+    const int size3 = 7;
+    std::cout << "\nTest 3- as std::vector<float>...";
+    std::vector<float> arr3 = {6.56, 12.62, 200.78, 768.484, 19.27, 68.87, 9.6};
+    sorting::pancakeSort(arr3, size3);
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:101]:
```suggestion
    sorting::pancake_sort::pancakeSort(arr3, size3);
```

File: sorting/pancake_sort.cpp [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    std::cout << std::endl;
+}
+/**
+ * Our main function with example of sort method.
+ */
+int main() {
+    test();
+    return 0;
+}
+
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:116]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/divyansh12323/C-Plus-Plus/actions

File: sorting/pancake_sort.cpp [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+            arr[start] = arr[end];
+            arr[end] = temp;
+            start++;
+            end--;
+        }
+    }
+    /**
+    * @brief This implementation is for a C-style array input that gets modified in place.
+    * @param [start,end] arr our vector of elements.
+    * @param size size of given array
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:55]: What does this function return? Document it here.
```suggestion
    * @param size size of given array
    * @returns description
```

File: sorting/pancake_sort.cpp [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+/**
+ * @namespace pancake_sort
+ * @brief Functions for [Pancake sort](https://en.wikipedia.org/wiki/Pancake_sorting) algorithm
+ */
+namespace pancake_sort {
+    /**
+    * @brief This implementation is for reversing elements in a a C-style array .
+    * @param [start,end] arr our vector of elements.
+    * @param start starting index of array
+    * @param end ending index of array
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:39]:
```suggestion
    * @param end ending index of array
    * @returns void
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/divyansh12323/C-Plus-Plus/actions

**Comment by divyansh12323** [comment]
I have made the requested changes @Panquesito7 . As for the  github workflows should i wait for the PR to fix that?

File: sorting/pancake_sort.cpp [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+    }
+}  // namespace pancake_sort
+}  // namespace sorting
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
+
```
Comment (Panquesito7) [inline sorting/pancake_sort.cpp:86]:
```suggestion
```

**Review by Panquesito7 — APPROVED** [review]
LGTM, good work. 👍 😄

**Comment by Panquesito7** [comment]
> GitHub workflows should I wait for the PR to fix that?

There's nothing to do for now; the workflows' code is good.
Seems to be a problem itself with GitHub Actions, not the workflows.

**Comment by divyansh12323** [comment]
So should i wait until this problem fixes @Panquesito7

**Review by rishabh-997 — APPROVED** [review]
Quite insightful suggestions @Panquesito7. LGTM too. Should we wait for GitHub actions to be fixed or merge this one as the PRs are piling up recently...

**Comment by Panquesito7** [comment]
> Should we wait for GitHub actions to be fixed

https://github.com/TheAlgorithms/C-Plus-Plus/pull/1189#issuecomment-707950148

**Comment by Panquesito7** [comment]
GitHub Actions seems to be running well now. 🙂

**Comment by divyansh12323** [comment]
Yes!! so can you merge the PR now @Panquesito7

**Review by kvedala — APPROVED** [review]
LGTM 👍

**Pull request merged**